### ISSUE TITLE
Define auxiliary data structures for type inference

### ISSUE THREAD

**Issue by jlapacik** [influxdata/flux#1953]
Define the following data structures for type inference:

* Type Environment
A type environment is a map of in scope variable names to their assigned polytype. Inference will add a new entry to the environment for each variable assignment. It will also perform a lookup for each identifier expression that it encounters during the inference process.

* Substitution
A substitution is map of type variables to equivalent monotypes. A substitution is produced when unifying two monotypes. Substitutions are applied to monotypes, polytypes, and type environments. A valid substitution is one in which no key (or type variable) in the substitution is referenced by any of the values in the substitution. A substitution must always preserve this invariant. And finally substitutions can and will be merged together during inference, eventually producing a substitution for the entire program.

* Constraints
There are two types of constraints, equality and kind constraints. An equality constraint asserts that two monotypes are equal and will be unified at some point, producing a substitution. A kind constraint asserts that a monotype belongs to a certain type class. 

* Constraint Map
A constraint map is a map of type variables to kinds. A type variable that is present in the constraint map during inference must not be present as a key in the global substitution. In the case that a type variable is assigned to a monotype in the global substitution, that type variable must be removed from the constraint map if present.

**Comment by affo** [comment]
@jlapacik 

> A valid substitution is one in which no key (or type variable) in the substitution is referenced by any of the values in the substitution

Is this to avoid cyclic substitutions by construction?

Say I have

```
t0 -> Int
```

And then I find

```
t0 + t1
```

This would add a contraint like 

```
t1 -> t0
```

But it cannot be in the substitution like that, because we would violate the invariant, so, we should directly "expand" the substitution:

```
t0 -> Int
t1 -> Int
```

How far I got from something reasonable? 😅

**Comment by affo** [comment]
@jlapacik 

>  A type variable that is present in the constraint map during inference must not be present as a key in the global substitution. In the case that a type variable is assigned to a monotype in the global substitution, that type variable must be removed from the constraint map if present.

Let me rephrase to understand if I got it. You are saying that once you have a valid substitution for a type variable that matches some kind, then there is no sense in keeping a constraint on that type variable. That would only make the constraint map heavier.

**Comment by jlapacik** [comment]
@affo you are correct in both cases. For substitutions, an equivalent definition is that substitutions must be idempotent. That is, for any substitution `s` and any type `x`, `s(x) = s(s(x))`.

For your second question, substitutions define an equality relation between types whereas the constraint map defines containment relations. Therefore once a type variable is assigned inside the substitution, we know its exact type. Any information pertaining to that type variable inside the constraint map becomes redundant at that point.

**Comment by jlapacik** [comment]
Closed by #2061.
